lib/net/ssh/proxy/command.rb

```diff
@@ -56,13 +56,17 @@ def open(host, port, connection_options = nil)
       }
       begin
         io = IO.popen(command_line, "r+")
-        if result = Net::SSH::Compat.io_select([io], nil, [io], 60)
-          if result.last.any? || io.eof?
-            io.close
-            raise "command failed"
+        begin
+          if result = Net::SSH::Compat.io_select([io], nil, [io], 60)
+            if result.last.any? || io.eof?
+              raise "command failed"
+            end
+          else
+            raise "command timed out"
           end
-        else
-          raise "command timed out"
+        rescue
+          io.close
```

Comment:
Yes it looks a bit cleaner thanks

> There is still the issue that io.close does not immediately kill the pipe, and so it can hang around for quite a bit after the timeout happens (2 minutes on my machine).

it should solve the issue of the process left behind as a zombie won't it? Or does `io.close` itself hangs? (blocking the net-ssh process)

Killing the process on timeout sounds strange for first.

> Also, this is being a bit pedantic, but this doesn't handle the case of the parent dying - in that case, there will again be orphaned children.

By `parent` you mean the ruby process running the net-ssh code?

Reply:
Yes, I mean that `io.close` itself will hang, blocking the net-ssh process. If you have a look [here](http://man7.org/linux/man-pages/man3/popen.3.html), it states that pclose() will `wait for the associated process to terminate and return the exit status of the command as returned by wait4(2)`.

So in this version of the code, if a timeout/exception happens and `io.close` is called, it will wait until the process itself terminates unless we send it a signal to terminate now - for example a ProxyCommand like `sleep 300` will make `io.close` hang for 300 seconds.

Yes, I meant the ruby process running the net-ssh code by `parent` :)

Reply:
So on the first issue, we can:

1. ignore it and let net-ssh block in close
2. send a kill signal as you suggested `Process.kill('TERM', io.pid) ; io.close`
3. instead of `io.close` do `Thread.new { io.close }`

i'm not sure which one is the best. Blocking `net-ssh` is not ideal, but nor are the other solutions. Maybe let's go with 2?

In case `net-ssh` terminates we won't leave zombies around, just the command started by `IO.popen` might continue to run. I'd say it's reasonable, probably happens with raw ssh as well. 

```sh
 ssh -o ProxyCommand='sleep 311' remote_server1
 kill -9 <ssh's pid>
 # sleep is still running
```

Reply:
I'm going to give this some more thought. I'm also in favour of approach 2, but I had a look at the [openssh source code](https://github.com/openssh/openssh-portable/blob/2937dd02c572a12f33d5c334d518f6cbe0b645eb/sshconnect.c#L270) and their approach seems to be sending a SIGHUP signal to the proxy command to clean up, but they they do not wait on it since it may hang - `io.close` would wait on the process though.

Reply:
@DimitriosLisenko  any update on this?